### PR TITLE
Implementing CNN as unsupervised learning using FFA

### PR DESCRIPTION
**Pull request by cyclotomicextension**
Hello, I tried creating CNNBuildOperation, ConvBNReLU, ConvFFLayer and CNNForwardForwardTrainer where I trained CNN via KMeans clustering. Please let me know what you think!

### CODE REVIEW

**Review by diegofiori — COMMENTED** [review]
Hi @cyclotomicextension, great to see your implementation, thank you! Please take a look at our [contribution guidelines](https://github.com/nebuly-ai/nebullvm/blob/main/CONTRIBUTING.md) for information about the style and code format we are using in the library. I left a couple of comments on the PR with some ideas that I think would be nice to add to the implementation

File: apps/accelerate/forward_forward/forward_forward/api/ffa.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from forward_forward import train_with_forward_forward_algorithm
```
[the remark this answers is not in the document]
Comment (cyclotomicextension) [reply apps/accelerate/forward_forward/forward_forward/api/ffa.py:1]: Yes, I am sorry, this was just for myself.

File: apps/accelerate/forward_forward/forward_forward/operations/trainers.py [hunk 2]
```diff
@@ -183,3 +188,53 @@ def _train(
...
+            kmeans = KMeans(n_clusters=10, random_state=0)
+            kmeans.fit(x_train)
+
+            # Predict the cluster assignments for each data point
+            train_clusters = kmeans.predict(x_train)
+            test_clusters = kmeans.predict(x_test)
+
+            for j, (data, target) in enumerate(self.train_data):
+                # TODO: THE IMAGE SHAPE SHOULD NOT BE DEFINED HERE
+                data = data.to(device).reshape(-1, 28 * 28)
```
[the remark this answers is not in the document]
Comment (cyclotomicextension) [reply apps/accelerate/forward_forward/forward_forward/operations/trainers.py:213]: Oh right sorry! I will fix this.

File: apps/accelerate/forward_forward/forward_forward/api/functions.py [hunk 3]
```diff
@@ -27,6 +27,9 @@ def train_with_forward_forward_algorithm(
     elif model_type is ForwardForwardModelType.RECURRENT:
         input_size = 28 * 28
         output_size = len(datasets.MNIST.classes)
+    elif model_type is ForwardForwardModelType.CNN:
+        input_size = 28 * 28 * 1
```
[the remark this answers is not in the document]
Comment (cyclotomicextension) [reply apps/accelerate/forward_forward/forward_forward/api/functions.py:31]: Yes right, the input should be a 2D image and not a flattened 1D image. I will change the input_size to (28, 28, 1).

File: apps/accelerate/forward_forward/forward_forward/operations/trainers.py [hunk 4]
```diff
@@ -183,3 +188,53 @@ def _train(
...
+        for epoch in range(epochs):
+            accumulated_goodness = None
+            model.train()
+            # Preprocess the data
+            x_train = [data.flatten().numpy() for data, label in self.train_data]
+            x_test = [data.flatten().numpy() for data, label in self.test_data]
+
+            # Perform clustering
+            kmeans = KMeans(n_clusters=10, random_state=0)
+            kmeans.fit(x_train)
```
[the remark this answers is not in the document]
Comment (cyclotomicextension) [reply apps/accelerate/forward_forward/forward_forward/operations/trainers.py:205]: I see, I shall do this now. Thank you for your input! So I will just add 
`kmeans = KMeans(n_clusters=10, random_state=0)
train_labels = kmeans.fit_predict(x_train)
train_dataset = torch.utils.data.TensorDataset(x_train, train_labels)
train_dataloader = torch.utils.data.DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
for j, (data, target) in enumerate(train_dataloader):
data = data.to(device)
target =  target = torch.functional.F.one_hot(torch.tensor([target]),
                    num_classes=10,
                ).to(device)`
Is this correct?

**Review by cyclotomicextension — COMMENTED** [review]
I have fixed all the issues pointed out earlier. Thanks!

File: apps/accelerate/forward_forward/forward_forward/operations/trainers.py [hunk 5]
```diff
@@ -183,3 +188,49 @@ def _train(
             predictions, _ = model.positive_eval(test_data, theta)
             perplexity = compute_perplexity(predictions)
             self.logger.info(f"Perplexity: {perplexity}")
+
+            
+class CNNForwardForwardTrainer(BaseForwardForwardTrainer):
```
Comment (diegofiori) [inline apps/accelerate/forward_forward/forward_forward/operations/trainers.py:193]: what about renaming it as `UnsupervisedCNNForwardForwardTrainer`?

File: apps/accelerate/forward_forward/forward_forward/operations/trainers.py [hunk 6]
```diff
@@ -183,3 +188,49 @@ def _train(
...
+    def _train(self, epochs: int, theta: float, device: str, **kwargs):
+        model = self.model.to(device)
+        for epoch in range(epochs):
+            accumulated_goodness = None
+            model.train()
+            # Preprocess the data
+            x_train = [data.flatten().numpy() for data, label in self.train_data]
```
Comment (diegofiori) [inline apps/accelerate/forward_forward/forward_forward/operations/trainers.py:200]: why are you flattening the input tensors?

Comment (cyclotomicextension) [reply apps/accelerate/forward_forward/forward_forward/operations/trainers.py]: I do not understand this. So, I get that I should not flatten x_train. Since, x_train is used in training so this means its better to have x_test as unflattened tensor too, right?

Comment (diegofiori) [reply apps/accelerate/forward_forward/forward_forward/operations/trainers.py]: Here the problem is that the network expects as input not-flattened data, while KMeans expects as input flattened data. Furthermore, we should considered that in forward-forward we should add the negative-examples too for training the model. 

I propose to modify the code in the following way: 
* Instead of creating a new dataloader using the `x_train` numpy data, we use KMeans for obtaining the "unsupervised labels" in a new function `get_unsupervised_label` and then we use the obtained labels to get both the positive and negative data. In the example below the function directly returns the dataloader we should use in the training loop (and the kmeans model).
```python
def get_unsupervised_label(self, device):
    x_train = np.concatenate(
        [data.detach().cpu().numpy() for data, label in
         self.train_data], axis=0)

    # Perform clustering
    kmeans = KMeans(n_clusters=10, random_state=0)
    train_labels = kmeans.fit_predict(x_train)
    train_labels = torch.from_numpy(train_labels).to(device)
    label_injector = LabelsInjector([f"cluster_{i}" for i in range(10)])

    train_bs = self.train_data.batch_size
    progressive_train_dataset = ProgressiveTrainingDataset(
        (label_injector.inject_train(
            x.unsqueeze(1), 
            train_labels[i*train_bs: (i+1)*train_bs]
        ) for i, (x, y) in enumerate(self.train_data))
    )
    progressive_train_dataloader = torch.utils.data.DataLoader(
        progressive_train_dataset,
        batch_size=2 * train_bs, shuffle=False
    )
    return progressive_train_dataloader, kmeans
```
* When computing the validation metrics we can compute the kmeans prediction on the fly and then use the generated label for computing accuracy. I expect the validation loop to be something like
```python
correct = 0
with torch.no_grad():
    for data, target in self.test_data:
        data = data.to(device)
        numpy_data = data.flatten().cpu().numpy()
        target = torch.from_numpy(kmeans.predict(numpy_data)).to(device)
        pred, _ = model.positive_eval(data.unsqueeze(1), theta)
        correct += pred.eq(target.view_as(pred)).sum().item()
self.logger.info(
    f"Test accuracy: {correct} / 10000 ({correct / 10000 * 100}%)"
)
```

What do you think?

Comment (cyclotomicextension) [reply apps/accelerate/forward_forward/forward_forward/operations/trainers.py]: I sort of understand what you are saying. I did something similar [here](https://github.com/cyclotomicextension/Contrastive-Learning-via-FFA/blob/main/ffa_implemention_of_contrastive_learning.py). I shall work on this now! Thanks a lot

**Comment by cyclotomicextension** [comment]
Hello, I had created a PR regarding hugging face, hence, had to close this. I will reopen this with the same title. Sorry for the inconvenience.
